File: 2022/day13.cpp
// https://adventofcode.com/2022/day/13

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <cctype>
#include <iostream>
#include <print>
#include <ranges>
#include <variant>

namespace stdv = std::views;
namespace stdr = std::ranges;

struct Packet;
using PacketList = Vec<Packet>;
using PacketPair = aoc::point_type<Packet>;
using Input = Vec<PacketPair>;

struct Packet {
  std::variant<u32, PacketList> data;
  constexpr std::strong_ordering operator<=>(Packet const& other) const;
  constexpr bool operator==(Packet const& other) const = default;

  friend std::ostream& operator<<(std::ostream& out, Packet const& p) {
    aoc::match(
        p.data, //
        [&](u32 num) { out << num; },
        [&](PacketList const& list) {
          out << '[';
          for (let& v : list) {
            out << v << ',';
          }
          out << ']';
        });
    return out;
  }
};

fn parse_list_inner(str s) -> Packet {
  auto list_stack = Vec<PacketList>{};
  list_stack.emplace_back();
  auto current_number = Option<u32>{};
  let push_number = [&] {
    if (current_number.has_value()) {
      list_stack.back().emplace_back(*current_number);
    }
    current_number = None;
  };
  for (let c : s) {
    if (std::isdigit(c)) {
      let digit = aoc::to_number<u32>(c);
      if (current_number.has_value()) {
        current_number = *current_number * 10 + digit;
      } else {
        current_number = digit;
      }
    } else {
      switch (c) {
        case '[': {
          push_number();
          list_stack.emplace_back();
        } break;
        case ']': {
          push_number();
          let last = aoc::pop_stack(list_stack);
          list_stack.back().emplace_back(std::move(last));
        } break;
        case ',': {
          push_number();
        } break;
        default:
          AOC_ASSERT(std::isdigit(c), "Invalid character in input");
      }
    }
  }
  push_number();
  return Packet{list_stack.back()};
}

auto parse(String const& filename) -> Input {
  return aoc::split(aoc::trim(aoc::read_file(filename)), "\n\n") |
         stdv::transform([](str pair) {
           let[first, second] = aoc::split_once(pair, '\n');
           return PacketPair{
               parse_list_inner(first.substr(1, first.size() - 2)),
               parse_list_inner(second.substr(1, second.size() - 2)),
           };
         }) |
         aoc::ranges::to<Input>();
}

constexpr std::strong_ordering Packet::operator<=>(Packet const& other) const {
  let overload = aoc::overload{
      [](u32 left, u32 right) { return left <=> right; }, // FORCE FORMATTING
      [](PacketList const& left, PacketList const& right) {
        return std::lexicographical_compare_three_way(
            std::begin(left), std::end(left), //
            std::begin(right), std::end(right));
      },
      [](u32 left, PacketList const& right) {
        return PacketList(1, Packet{left}) <=> right;
      },
      [](PacketList const& left, u32 right) {
        return left <=> PacketList(1, Packet{right});
      }};
  return std::visit(overload, data, other.data);
}

fn solve_case1(Input const& input) -> u32 {
  return aoc::ranges::accumulate(
      input | stdv::enumerate | stdv::filter([](let& index_val_pair) {
        let & [ first, second ] = std::get<1>(index_val_pair);
        return first < second;
      }) | stdv::transform([](let& index_val_pair) {
        return std::get<0>(index_val_pair) + 1;
      }),
      u32{});
}

fn solve_case2(Input const& input) -> u32 {
  auto packets = input | stdv::join | aoc::ranges::to<Vec<Packet>>();
  let p2 = parse_list_inner("[2]");
  let p6 = parse_list_inner("[6]");
  packets.push_back(p2);
  packets.push_back(p6);
  stdr::sort(packets);
  return (aoc::ranges::position(packets, p2).value() + 1) *
         (aoc::ranges::position(packets, p6).value() + 1);
}

int main() {
  std::println("Part 1");
  let example = parse("day13.example");
  AOC_EXPECT_RESULT(13, solve_case1(example));
  let input = parse("day13.input");
  AOC_EXPECT_RESULT(5198, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(140, solve_case2(example));
  AOC_EXPECT_RESULT(22344, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2022/src/day13.rs
use itertools::Itertools;
use std::cmp::Ordering;

type PacketList = Vec<Packet>;

#[derive(Clone, Debug, PartialEq, Eq)]
enum Packet {
    Integer(u32),
    List(PacketList),
}
impl Ord for Packet {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Packet::Integer(left), Packet::Integer(right)) => left.cmp(right),
            (Packet::List(left), Packet::List(right)) => left
                .iter()
                .zip(right)
                .fold(Ordering::Equal, |acc, (l, r)| acc.then(l.cmp(r)))
                .then(left.len().cmp(&right.len())),
            (Packet::Integer(left), Packet::List(_)) => {
                Packet::List(PacketList::from([Packet::Integer(*left)])).cmp(other)
            }
            (Packet::List(_), Packet::Integer(right)) => {
                self.cmp(&Packet::List(PacketList::from([Packet::Integer(*right)])))
            }
        }
    }
}
impl PartialOrd for Packet {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

type Input = Vec<(Packet, Packet)>;

fn parse_list_inner(s: &str) -> Packet {
    let mut list_stack = Vec::<PacketList>::new();
    list_stack.push(PacketList::new());
    let mut current_number = None;
    let push_number = |list_stack: &mut PacketList, current_number: &mut Option<u32>| {
        if let Some(num) = *current_number {
            list_stack.push(Packet::Integer(num));
        }
        *current_number = None;
    };
    for c in s.chars() {
        match c {
            '[' => {
                push_number(&mut list_stack.last_mut().unwrap(), &mut current_number);
                list_stack.push(PacketList::new());
            }
            ']' => {
                push_number(&mut list_stack.last_mut().unwrap(), &mut current_number);
                let last = list_stack.pop().unwrap();
                list_stack.last_mut().unwrap().push(Packet::List(last));
            }
            ',' => {
                push_number(&mut list_stack.last_mut().unwrap(), &mut current_number);
            }
            digit if (digit.is_digit(10)) => {
                let digit = digit.to_digit(10).unwrap();
                if let Some(num) = &mut current_number {
                    *num = *num * 10 + digit;
                } else {
                    current_number = Some(digit);
                }
            }
            _ => unreachable!("Invalid character in input"),
        };
    }
    push_number(&mut list_stack.last_mut().unwrap(), &mut current_number);
    return Packet::List(list_stack.last_mut().unwrap().to_vec());
}

fn parse(filename: &str) -> Input {
    aoc::file::read_string(filename)
        .trim()
        .split("\n\n")
        .map(|pair| {
            let (first, second) = pair.split_once('\n').unwrap();
            return (
                parse_list_inner(&first[1..first.len() - 1]),
                parse_list_inner(&second[1..second.len() - 1]),
            );
        })
        .collect()
}

fn solve_case1(input: &Input) -> usize {
    input
        .iter()
        .enumerate()
        .filter(|(_, (first, second))| first < second)
        .map(|(index, _)| index + 1)
        .sum()
}

fn solve_case2(input: &Input) -> usize {
    let mut packets = input
        .iter()
        .flat_map(|(first, second)| [first, second])
        .cloned()
        .collect_vec();
    let p2 = parse_list_inner("[2]");
    let p6 = parse_list_inner("[6]");
    packets.push(p2.clone());
    packets.push(p6.clone());
    packets.sort();
    return (packets.iter().position(|p| *p == p2).unwrap() + 1)
        * (packets.iter().position(|p| *p == p6).unwrap() + 1);
}

fn main() {
    println!("Part 1");
    let example = parse("day13.example");
    assert_eq!(13, solve_case1(&example));
    let input = parse("day13.input");
    assert_eq!(5198, solve_case1(&input));

    println!("Part 2");
    assert_eq!(140, solve_case2(&example));
    assert_eq!(22344, solve_case2(&input));
}
